lib/spack/spack/directives.py

```diff
@@ -598,6 +601,64 @@ def _execute_depends_on(pkg: "spack.package_base.PackageBase"):
     return _execute_depends_on
 
 
+#: Store whether a given Spec source/binary should not be redistributed.
+class DisableRedistribute:
+    def __init__(self, source, binary):
+        self.source = source
+        self.binary = binary
+
+
+@directive("disable_redistribute")
+def redistribute(source=None, binary=None, when: WhenType = None):
+    """Can be used inside a Package definition to declare that
+    the package source and/or compiled binaries should not be
+    redistributed.
+
+    By default, Packages allow source/binary distribution (i.e. in
+    mirrors). Because of this, and because overlapping enable/
+    disable specs are not allowed, this directive only allows users
+    to explicitly disable redistribution for specs.
+    """
+
+    return lambda pkg: _execute_redistribute(pkg, source, binary, when)
+
+
+def _execute_redistribute(
+    pkg: "spack.package_base.PackageBase", source=None, binary=None, when: WhenType = None
+):
+    if source is None and binary is None:
+        return
+    elif (source is True) or (binary is True):
+        raise DirectiveError(
+            "Source/binary distribution are true by default, they can only "
+            "be explicitly disabled."
+        )
```

Comment:
Why couldn't you have:

```python
class BuiltinPackage:
    redistribute(source=False, binary=False)
```

and:

```python
class MyRepoPackage(BuiltinPackage):
    redistribute(source=True, binary=True). # override setting in superclass
```

Even within `builtin`, I could see people wanting to create derived classes depending on a license change (e.g., for [Valkey](https://www.linuxfoundation.org/press/linux-foundation-launches-open-source-valkey-community) or [OpenTofu](https://www.linuxfoundation.org/press/announcing-opentofu), which are open source forks of recently-made-closed-source products).

Maybe I'm not completely seeing what the "only allow setting to false" thing buys us here.

Reply:
Within a single package, I think we agreed that we don't want to allow competing overlapping specs like

```
redistribute(source=False, when=@2:3)
redistribute(source=True, when=@2.5)
```

given that we default to `source/binary=True`, only `False` should be allowed (`True` would only ever be needed for creating overlaps). 

You are highlighting a case of 100% overlap (i.e. key-replacement); while I agree that should be doable, for subclasses, we (will) have `drop` (which is why this info was reorganized into a Spec-keyed dictionary): I think it's not worth complicating the single-class case to offer a limited version of what `drop` will achieve in a more-straightforward manner.